src/ynEIGEN/AssetRegistry.sol

```diff
@@ -0,0 +1,323 @@
+    bytes32 public constant UNPAUSER_ROLE = keccak256("UNPAUSER_ROLE");
+    
+    //--------------------------------------------------------------------------------------
+    //----------------------------------  VARIABLES  ---------------------------------------
+    //--------------------------------------------------------------------------------------
+
+     /// @notice List of supported ERC20 asset contracts.
+    IERC20[] public assets;
+
+    mapping(IERC20 => AssetData) public _assetData;
+
+    bool public actionsPaused;
+
+    IRateProvider public rateProvider;
+    IEigenStrategyManager public eigenStrategyManager;
+    IynEigen ynEigen;
+
+
+    //--------------------------------------------------------------------------------------
+    //----------------------------------  INITIALIZATION  ----------------------------------
+    //--------------------------------------------------------------------------------------
+
+    constructor() {
+       _disableInitializers();
+    }
+
+    struct Init {
+        IERC20[] assets;
+        IRateProvider rateProvider;
+        IEigenStrategyManager eigenStrategyManager;
+        IynEigen ynEigen;
+        address admin;
+        address pauser;
+        address unpauser;
+    }
+
+    function initialize(Init calldata init)
+        public
+        notZeroAddress(address(init.rateProvider))
+        notZeroAddress(address(init.admin))
+        notZeroAddress(address(init.pauser))
+        notZeroAddress(address(init.unpauser))
+        initializer {
+        __AccessControl_init();
+        __ReentrancyGuard_init();
+
+        _grantRole(DEFAULT_ADMIN_ROLE, init.admin);
+        _grantRole(PAUSER_ROLE, init.pauser);
+        _grantRole(UNPAUSER_ROLE, init.unpauser);
+
+        for (uint256 i = 0; i < init.assets.length; i++) {
+            if (address(init.assets[i]) == address(0)) {
+                revert ZeroAddress();
+            }
+            assets.push(init.assets[i]);
+            _assetData[init.assets[i]] = AssetData({
+                active: true
+            });
+        }
+
+        rateProvider = init.rateProvider;
+        eigenStrategyManager = init.eigenStrategyManager;
+        ynEigen = init.ynEigen;
+    }
+
+    //--------------------------------------------------------------------------------------
+    //----------------------------------  ASSET MANAGEMENT  --------------------------------
+    //--------------------------------------------------------------------------------------
+
+    /**
+     * @notice Adds a new asset to the system.
+     * @dev Adds an asset to the _assetData mapping and sets it as active. This function can only be called by the strategy manager.
+     * @param asset The address of the ERC20 token to be added.
+     */
+    function addAsset(IERC20 asset) public onlyRole(ASSET_MANAGER_ROLE) whenNotPaused {
+        require(!_assetData[asset].active, "Asset already active");
```

Comment:
not sure if I like this or not, but we could check: `require(address(asset) != address(0), "Invalid asset address");` but given this is an ops role, and that we can remove assets, it might not be important should an error be made.

Reply:
added the nonZeroAddress checks, i agree they are good